In [2]:
import geocoder
import requests
import pandas as pd
import numpy as np
import time

pd.options.display.float_format = '{:,.6f}'.format

In [3]:
from geopy.geocoders import Nominatim
nom=Nominatim()

ModuleNotFoundError: No module named 'geopy'

Client.coordinates('Україна, Черкаси, Школа №10')

g = geocoder.yandex('Черкаси, ДНЗ №41')
g.json

geocoder.yandex('Україна, Черкаси, Школа №10').lat

g = geocoder.google('черкаси школа №31')
g.wkt

### Тест якоїсь херні

In [30]:
df_map_pivot = pd.read_csv('24_regions.csv', index_col = 0, encoding='UTF-8')
df_map_pivot

Заклад    Витрачена сума  \
0      Івано-Франківськ ДНЗ №1    743,150.000000   
1     Івано-Франківськ ДНЗ №10  1,430,492.050000   
2     Івано-Франківськ ДНЗ №11  1,140,318.000000   
3     Івано-Франківськ ДНЗ №12    370,266.600000   
4     Івано-Франківськ ДНЗ №14    852,164.710000   
5     Івано-Франківськ ДНЗ №15    623,348.000000   
6     Івано-Франківськ ДНЗ №16  1,165,309.000000   
7     Івано-Франківськ ДНЗ №17     96,200.000000   
8     Івано-Франківськ ДНЗ №18  2,197,095.900000   
9     Івано-Франківськ ДНЗ №19    390,838.790000   
10     Івано-Франківськ ДНЗ №2  2,924,688.360000   
11    Івано-Франківськ ДНЗ №20    714,286.160000   
12    Івано-Франківськ ДНЗ №22    493,549.470000   
13    Івано-Франківськ ДНЗ №23    602,035.000000   
14    Івано-Франківськ ДНЗ №27    871,600.000000   
15    Івано-Франківськ ДНЗ №28  1,150,781.530000   
16    Івано-Франківськ ДНЗ №29    443,671.200000   
17     Івано-Франківськ ДНЗ №3  1,757,670.030000   
18    Івано-Франківськ ДНЗ №30    572,032.410000   
19    Івано-Франківськ ДНЗ №33    755,256.830000   
20    Івано-Франківськ ДНЗ №34  1,114,906.000000   
21    Івано-Франківськ ДНЗ №35    282,331.000000   
22    Івано-Франківськ ДНЗ №36    624,217.400000   
23     Івано-Франківськ ДНЗ №4    360,377.000000   
24     Івано-Франківськ ДНЗ №6  1,958,932.000000   
25     Івано-Франківськ ДНЗ №7    242,114.370000   
26     Івано-Франківськ ДНЗ №9    307,834.000000   
27    Івано-Франківськ ДЮСШ №2    468,555.990000   
28    Івано-Франківськ ДЮСШ №3    489,598.000000   
29     Івано-Франківськ Каскад 63,316,748.000000   
...                        ...               ...   
3924        Чернігів Школа №16  4,238,478.410000   
3925        Чернігів Школа №17    440,119.300000   
3926        Чернігів Школа №18  1,022,585.810000   
3927        Чернігів Школа №19 13,231,003.250000   
3928         Чернігів Школа №2  4,776,413.990000   
3929        Чернігів Школа №20  2,116,146.820000   
3930        Чернігів Школа №21    577,455.060000   
3931        Чернігів Школа №22  1,396,732.760000   
3932        Чернігів Школа №23    296,031.540000   
3933        Чернігів Школа №24  1,913,574.480000   
3934        Чернігів Школа №25  1,327,684.940000   
3935        Чернігів Школа №27  1,435,366.790000   
3936        Чернігів Школа №28  1,855,405.780000   
3937        Чернігів Школа №29  2,013,545.140000   
3938         Чернігів Школа №3 34,424,701.200000   
3939        Чернігів Школа №30  2,479,819.690000   
3940        Чернігів Школа №31  3,224,386.420000   
3941        Чернігів Школа №32    948,584.000000   
3942        Чернігів Школа №33  2,111,841.530000   
3943        Чернігів Школа №34  7,270,740.220000   
3944        Чернігів Школа №35 14,709,825.840000   
3945        Чернігів Школа №36  2,637,477.100000   
3946        Чернігів Школа №38    461,041.440000   
3947         Чернігів Школа №4    535,147.710000   
3948        Чернігів Школа №46  1,083,336.000000   
3949         Чернігів Школа №5  2,236,865.250000   
3950         Чернігів Школа №6  5,457,437.260000   
3951         Чернігів Школа №7  3,883,770.380000   
3952         Чернігів Школа №9  1,373,056.390000   
3953          Чернігів Шкільне    307,710.580000   

                                              Посилання  \
0     https://dozorro.org/tender/UA-2018-07-03-00170...   
1     https://dozorro.org/tender/UA-2017-08-09-00125...   
2     https://dozorro.org/tender/UA-2017-07-13-00164...   
3     https://dozorro.org/tender/UA-2018-07-24-00047...   
4     https://dozorro.org/tender/UA-2017-08-01-00225...   
5     https://dozorro.org/tender/UA-2017-09-27-00034...   
6     https://dozorro.org/tender/UA-2017-11-23-00349...   
7     https://dozorro.org/tender/UA-2019-03-29-000986-b   
8     https://dozorro.org/tender/UA-2018-08-08-00096...   
9     https://dozorro.org/tender/UA-2017-07-20-00155...   
10    https://dozorro.org/tender/UA-2017-10-05-00039...   
11    https://dozorro.org/tender/UA-2017-03-31-00112...   
12    https://dozorro.org/tender

In [15]:
from geopy.extra.rate_limiter import RateLimiter
from geopy.geocoders import Nominatim
geolocator = Nominatim(user_agent="specify_your_app_name_here")
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=1)

In [31]:
test1 = df_map_pivot[50:80].copy()
test1["Coordinates"]=test1["address"].apply(geocode)

In [32]:
print(test1['Coordinates'][1])
test1

None


Заклад    Витрачена сума  \
0    Івано-Франківськ ДНЗ №1    743,150.000000   
1   Івано-Франківськ ДНЗ №10  1,430,492.050000   
2   Івано-Франківськ ДНЗ №11  1,140,318.000000   
3   Івано-Франківськ ДНЗ №12    370,266.600000   
4   Івано-Франківськ ДНЗ №14    852,164.710000   
5   Івано-Франківськ ДНЗ №15    623,348.000000   
6   Івано-Франківськ ДНЗ №16  1,165,309.000000   
7   Івано-Франківськ ДНЗ №17     96,200.000000   
8   Івано-Франківськ ДНЗ №18  2,197,095.900000   
9   Івано-Франківськ ДНЗ №19    390,838.790000   
10   Івано-Франківськ ДНЗ №2  2,924,688.360000   
11  Івано-Франківськ ДНЗ №20    714,286.160000   
12  Івано-Франківськ ДНЗ №22    493,549.470000   
13  Івано-Франківськ ДНЗ №23    602,035.000000   
14  Івано-Франківськ ДНЗ №27    871,600.000000   
15  Івано-Франківськ ДНЗ №28  1,150,781.530000   
16  Івано-Франківськ ДНЗ №29    443,671.200000   
17   Івано-Франківськ ДНЗ №3  1,757,670.030000   
18  Івано-Франківськ ДНЗ №30    572,032.410000   
19  Івано-Франківськ ДНЗ №33    755,256.830000   
20  Івано-Франківськ ДНЗ №34  1,114,906.000000   
21  Івано-Франківськ ДНЗ №35    282,331.000000   
22  Івано-Франківськ ДНЗ №36    624,217.400000   
23   Івано-Франківськ ДНЗ №4    360,377.000000   
24   Івано-Франківськ ДНЗ №6  1,958,932.000000   
25   Івано-Франківськ ДНЗ №7    242,114.370000   
26   Івано-Франківськ ДНЗ №9    307,834.000000   
27  Івано-Франківськ ДЮСШ №2    468,555.990000   
28  Івано-Франківськ ДЮСШ №3    489,598.000000   
29   Івано-Франківськ Каскад 63,316,748.000000   
..                       ...               ...   
70           Вінниця ДНЗ №19     66,300.000000   
71            Вінниця ДНЗ №2    154,176.000000   
72           Вінниця ДНЗ №21 13,124,992.200000   
73           Вінниця ДНЗ №23 19,355,632.000000   
74           Вінниця ДНЗ №24    766,481.040000   
75           Вінниця ДНЗ №25     67,200.000000   
76           Вінниця ДНЗ №26 17,263,012.650000   
77           Вінниця ДНЗ №27    730,968.960000   
78           Вінниця ДНЗ №28    296,618.800000   
79           Вінниця ДНЗ №29  1,213,470.880000   
80            Вінниця ДНЗ №3    270,827.000000   
81           Вінниця ДНЗ №30 10,485,735.750000   
82           Вінниця ДНЗ №31  2,624,662.020000   
83           Вінниця ДНЗ №34    154,675.000000   
84           Вінниця ДНЗ №35    365,705.000000   
85           Вінниця ДНЗ №37    324,662.000000   
86           Вінниця ДНЗ №38    347,930.000000   
87            Вінниця ДНЗ №4  5,105,676.920000   
88           Вінниця ДНЗ №42 10,776,288.620000   
89           Вінниця ДНЗ №43    221,089.020000   
90           Вінниця ДНЗ №45  1,460,008.720000   
91           Вінниця ДНЗ №46    736,828.020000   
92           Вінниця ДНЗ №47  6,024,531.000000   
93            Вінниця ДНЗ №5    164,675.340000   
94           Вінниця ДНЗ №50    144,150.000000   
95           Вінниця ДНЗ №51    484,658.290000   
96           Вінниця ДНЗ №52  9,277,679.590000   
97           Вінниця ДНЗ №57    637,177.040000   
98           Вінниця ДНЗ №59    231,618.000000   
99            Вінниця ДНЗ №6    154,698.670000   

                                            Посилання  \
0   https://dozorro.org/tender/UA-2018-07-03-00170...   
1   https://dozorro.org/tender/UA-2017-08-09-00125...   
2   https://dozorro.org/tender/UA-2017-07-13-00164...   
3   https://dozorro.org/tender/UA-2018-07-24-00047...   
4   https://dozorro.org/tender/UA-2017-08-01-00225...   
5   https://dozorro.org/tender/UA-2017-09-27-00034...   
6   https://dozorro.org/tender/UA-2017-11-23-00349...   
7   https://dozorro.org/tender/UA-2019-03-29-000986-b   
8   https://dozorro.org/tender/UA-2018-08-08-00096...   
9   https://dozorro.org/tender/UA-2017-07-20-00155...   
10  https://dozorro.org/tender/UA-2017-10-05-00039...   
11  https://dozorro.org/tender/UA-2017-03-31-00112...   
12  https://dozorro.org/tender/UA-2018-07-24-00028...   
13  https://dozorro.org/tender/UA-2017-10-11-00023...   
14  https://dozorro.org/tender/UA-2017-10-31-00131...   
15  https

In [26]:
test1

Заклад   Витрачена сума  \
0   Івано-Франківськ ДНЗ №1   743,150.000000   
1  Івано-Франківськ ДНЗ №10 1,430,492.050000   
2  Івано-Франківськ ДНЗ №11 1,140,318.000000   
3  Івано-Франківськ ДНЗ №12   370,266.600000   
4  Івано-Франківськ ДНЗ №14   852,164.710000   
5  Івано-Франківськ ДНЗ №15   623,348.000000   
6  Івано-Франківськ ДНЗ №16 1,165,309.000000   
7  Івано-Франківськ ДНЗ №17    96,200.000000   
8  Івано-Франківськ ДНЗ №18 2,197,095.900000   
9  Івано-Франківськ ДНЗ №19   390,838.790000   

                                           Посилання  \
0  https://dozorro.org/tender/UA-2018-07-03-00170...   
1  https://dozorro.org/tender/UA-2017-08-09-00125...   
2  https://dozorro.org/tender/UA-2017-07-13-00164...   
3  https://dozorro.org/tender/UA-2018-07-24-00047...   
4  https://dozorro.org/tender/UA-2017-08-01-00225...   
5  https://dozorro.org/tender/UA-2017-09-27-00034...   
6  https://dozorro.org/tender/UA-2017-11-23-00349...   
7  https://dozorro.org/tender/UA-2019-03-29-000986-b   
8  https://dozorro.org/tender/UA-2018-08-08-00096...   
9  https://dozorro.org/tender/UA-2017-07-20-00155...   

                         Адреса 2  \
0           вулиця Вовчинецька, 9   
1     вулиця Івана Миколайчука, 5   
2  вулиця Академіка Сахарова, 32а   
3                   вул. Довга, 4   
4          вулиця М.Підгірянки,21   
5                 вулиця Нова, 19   
6       вулиця Короля Данила, 15а   
7         вулиця Вовчинецька, 210   
8         вул. Івана Павла ІІ, 14   
9           вулиця Галицька, 101а   

                                          Адреса 1  \
0           Івано-Франківськ вулиця Вовчинецька, 9   
1     Івано-Франківськ вулиця Івана Миколайчука, 5   
2  Івано-Франківськ вулиця Академіка Сахарова, 32а   
3                   Івано-Франківськ вул. Довга, 4   
4          Івано-Франківськ вулиця М.Підгірянки,21   
5                 Івано-Франківськ вулиця Нова, 19   
6       Івано-Франківськ вулиця Короля Данила, 15а   
7         Івано-Франківськ вулиця Вовчинецька, 210   
8         Івано-Франківськ вул. Івана Павла ІІ, 14   
9           Івано-Франківськ вулиця Галицька, 101а   

                                         Coordinates  
0  (у Триндичихи, 9, Вовчинецька вулиця, Гірка, І...  
1  (Миколайчука, Каскад, Івано-Франківськ, Івано-...  
2                                               None  
3                                               None  
4                                               None  
5                                               None  
6  (Сонечко, 15а, Короля Данила вулиця, Кант, Іва...  
7  (210, Вовчинецька вулиця, Гірка, Івано-Франків...  
8                                               None  
9  (101а, Галицька вулиця, Княгинин, Івано-Франкі...

In [ ]:
df_map_pivot["Coordinates"]=df_map_pivot["Заклад"].apply(geocode)

df_map_pivot["Latitude"]=df_map_pivot["Coordinates"].apply(lambda x: x.latitude if x != None else None)
df_map_pivot["Longitude"]=df_map_pivot["Coordinates"].apply(lambda x: x.longitude if x != None else None)

#df_map_pivot

## Geocode by name of school

In [135]:
df_map_no_error = df_map_pivot[df_map_pivot['Latitude'].notnull()]
df_map_no_error

Заклад    Витрачена сума  \
0                                   Черкаси, Берегиня  3,589,376.000000   
9                                    Черкаси, ДНЗ №39    357,443.600000   
11                                   Черкаси, ДНЗ №45    921,979.400000   
12                                   Черкаси, ДНЗ №57  1,117,106.960000   
13                                   Черкаси, ДНЗ №59  1,051,721.260000   
17                                   Черкаси, ДНЗ №81  1,233,499.200000   
18                                   Черкаси, ДНЗ №91  1,052,301.090000   
25           Черкаси, Черкаська дитяча школа мистецтв  3,185,948.880000   
26  Черкаси, Черкаський професійний автодорожній л...  1,508,684.330000   
27              Черкаси, Черкаський професійний ліцей  2,577,956.840000   
29                                 Черкаси, Школа №10  2,509,415.430000   
30                                 Черкаси, Школа №11  2,404,258.230000   
31                                 Черкаси, Школа №12  3,255,307.130000   
32                                 Черкаси, Школа №13  4,737,529.320000   
33                                 Черкаси, Школа №17  8,380,343.860000   
34                                 Черкаси, Школа №18  6,198,063.140000   
35                                 Черкаси, Школа №19    491,465.300000   
36                                  Черкаси, Школа №2  2,457,121.980000   
37                                 Черкаси, Школа №20  3,973,798.320000   
38                                 Черкаси, Школа №21  3,084,428.210000   
39                                 Черкаси, Школа №22  3,494,725.020000   
40                                 Черкаси, Школа №24  4,789,368.050000   
41                                 Черкаси, Школа №26  7,950,985.020000   
42                                 Черкаси, Школа №27  4,589,153.320000   
43                                 Черкаси, Школа №28    348,390.830000   
44                                  Черкаси, Школа №3  5,031,781.140000   
45                                 Черкаси, Школа №30  4,578,438.150000   
46                                 Черкаси, Школа №31  7,864,100.370000   
47                                 Черкаси, Школа №32  8,659,487.420000   
48                                 Черкаси, Школа №33  6,872,410.400000   
49                                 Черкаси, Школа №34  2,473,302.100000   
50                                  Черкаси, Школа №4  2,495,789.430000   
51                                  Черкаси, Школа №5  2,410,312.520000   
52                                  Черкаси, Школа №6  4,355,353.240000   
53                                  Черкаси, Школа №7  2,682,962.030000   
54                                  Черкаси, Школа №8  2,586,596.730000   
55                                  Черкаси, Школа №9 11,849,415.220000   

                            Адреса закладу  \
0                       вул. Хоменка, 14/1   
9                          Нижня Горова,56   
11                                       -   
12                          Толстого, 73/1   
13                                       -   
17                                       -   
18                              Смірнова,4   
25                             Гастелло, 3   
26                            Онопрієнка,4   
27                вул. Ярослава Галана, 13   
29                      просп. Хіміків, 36   
30                     вул. Надпільна, 291   
31              вул. Б. Хмельницького, 118   
32              Гетьмана Сагайдачного, 146   
33                      вул. Хрещатик, 218   
34              вулиця Нечуя Левицького 12   
35    ВУЛИЦЯ УНІВЕРСИТЕТСЬКА, будинок 62-А   
36                 вул. Самійла Кішки, 187   
37                            Бидгощська 3   
38              вул. Нижня Горова, буд. 82   
39                 вул. Вербовецького, 108   
40                      вул. Бидгощська, 1   
41                   вул. Олени Теліги, 15   
42                         Сумгаїтська, 22   
43                        Героїв Майдану,3   


## Geocode Manual Edit

In [202]:
df_map_error = df_map_pivot[df_map_pivot['Latitude'].isnull()]

df_map_error = df_map_error.apply(lambda x: x.str.strip() if isinstance(x, str) else x).replace('-', np.nan)

df_map_error

df_map_error_noaddress = df_map_error[df_map_error['Адреса закладу'].isnull()]

df_map_error_noaddress.to_excel('CherkasyError_Edit.xls', index = True, header=True)

df_map_error_noaddress

Заклад   Витрачена сума Адреса закладу  \
2   Черкаси, ДНЗ №22   118,890.000000            NaN   
3   Черкаси, ДНЗ №25   197,411.000000            NaN   
4   Черкаси, ДНЗ №27   171,445.220000            NaN   
5   Черкаси, ДНЗ №29 2,473,554.000000            NaN   
6   Черкаси, ДНЗ №30   295,673.000000            NaN   
7   Черкаси, ДНЗ №37    68,901.000000            NaN   
8   Черкаси, ДНЗ №38   479,047.000000            NaN   
10  Черкаси, ДНЗ №41   572,465.000000            NaN   
14   Черкаси, ДНЗ №7   762,226.300000            NaN   
15  Черкаси, ДНЗ №70   912,793.340000            NaN   
16  Черкаси, ДНЗ №73 2,244,335.200000            NaN   

                                            Посилання Coordinates  Latitude  \
2    https://dozorro.org/tenderUA-2018-12-21-000170-b        None       nan   
3    https://dozorro.org/tenderUA-2018-12-21-000469-c        None       nan   
4    https://dozorro.org/tenderUA-2018-12-21-000300-b        None       nan   
5   https://dozorro.org/tenderUA-2018-09-13-000031...        None       nan   
6    https://dozorro.org/tenderUA-2018-11-05-001169-b        None       nan   
7    https://dozorro.org/tenderUA-2018-12-21-001029-a        None       nan   
8   https://dozorro.org/tenderUA-2018-12-22-001360...        None       nan   
10   https://dozorro.org/tenderUA-2018-12-14-000270-a        None       nan   
14   https://dozorro.org/tenderUA-2018-06-12-000620-b        None       nan   
15   https://dozorro.org/tenderUA-2018-08-16-000372-c        None       nan   
16  https://dozorro.org/tenderUA-2018-12-05-003706...        None       nan   

    Longitude  
2         nan  
3         nan  
4         nan  
5         nan  
6         nan  
7         nan  
8         nan  
10        nan  
14        nan  
15        nan  
16        nan

In [223]:
df_map_error_noaddress = pd.read_excel('CherkasyError_Edit.xls', encoding='UTF-8')
df_map_error_noaddress.to_csv('CherkasyError_Edit.csv', index = True, header=True)

df_map_error_noaddress.drop(df_map_error_noaddress.columns[df_map_error_noaddress.columns.str.contains('unnamed',case = False)],axis = 1, inplace = True)

df_map_error_noaddress

Заклад   Витрачена сума  Адреса закладу  \
0   Черкаси, ДНЗ №22   118,890.000000             nan   
1   Черкаси, ДНЗ №25   197,411.000000             nan   
2   Черкаси, ДНЗ №27   171,445.220000             nan   
3   Черкаси, ДНЗ №29 2,473,554.000000             nan   
4   Черкаси, ДНЗ №30   295,673.000000             nan   
5   Черкаси, ДНЗ №37    68,901.000000             nan   
6   Черкаси, ДНЗ №38   479,047.000000             nan   
7   Черкаси, ДНЗ №41   572,465.000000             nan   
8    Черкаси, ДНЗ №7   762,226.300000             nan   
9   Черкаси, ДНЗ №70   912,793.340000             nan   
10  Черкаси, ДНЗ №73 2,244,335.200000             nan   

                                            Посилання           Coordinates  \
0    https://dozorro.org/tenderUA-2018-12-21-000170-b  49,423767, 32,090116   
1    https://dozorro.org/tenderUA-2018-12-21-000469-c  49,435240, 31,999330   
2    https://dozorro.org/tenderUA-2018-12-21-000300-b  49,439276, 32,070601   
3   https://dozorro.org/tenderUA-2018-09-13-000031...  49,433309, 32,088953   
4    https://dozorro.org/tenderUA-2018-11-05-001169-b  49,433316, 32,088942   
5    https://dozorro.org/tenderUA-2018-12-21-001029-a  49,417642, 32,063832   
6   https://dozorro.org/tenderUA-2018-12-22-001360...  49,417983, 32,103502   
7    https://dozorro.org/tenderUA-2018-12-14-000270-a  49,430254, 32,051551   
8    https://dozorro.org/tenderUA-2018-06-12-000620-b  49,411104, 32,066168   
9    https://dozorro.org/tenderUA-2018-08-16-000372-c  49,427000, 32,013522   
10  https://dozorro.org/tenderUA-2018-12-05-003706...  49,427752, 32,012979   

    Latitude  Longitude  
0  49.423767  32.090116  
1  49.435240  31.999330  
2  49.439276  32.070601  
3  49.433309  32.088953  
4  49.433316  32.088942  
5  49.417642  32.063832  
6  49.417983  32.103502  
7  49.430254  32.051551  
8  49.411104  32.066168  
9  49.427000  32.013522  
10 49.427752  32.012979

## Geocode by address

In [218]:
df_map_error_address = df_map_error[df_map_error['Адреса закладу'].notnull()]

city = "Черкаси"
df_map_error_address["Адреса закладу"] = city + ", " + df_map_error_address["Адреса закладу"]

df_map_error_address

C:\Users\user1\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.


Заклад   Витрачена сума  \
1                                     Черкаси, Вулкан    68,316.000000   
19                                   Черкаси, ДЮСШ №1   141,594.200000   
20                          Черкаси, Музична школа №1 1,613,234.720000   
21                          Черкаси, Музична школа №3   581,440.000000   
22                     Черкаси, Перша міська гімназія 5,158,431.520000   
23  Черкаси, Центр дитячої та юнацької творчості м...   716,249.800000   
24  Черкаси, Черкаська дитяча художня школа ім. Да...   245,415.200000   
28  Черкаси, Черкаський фізико-математичний ліцей ... 3,066,886.090000   

                                      Адреса закладу  \
1                       Черкаси, вул. Ярославська, 6   
19                           Черкаси, Смілянська, 94   
20                       Черкаси, Б.Вишневецького,33   
21  Черкаси, ВУЛИЦЯ В'ЯЧЕСЛАВА ЧОРНОВОЛА, будинок 10   
22                   Черкаси, вул. Святотроїцька, 68   
23                 Черкаси, вул. Смілянська, буд. 33   
24                           Черкаси, Хрещатик,  214   
28                     Черкаси, вул. Благовісна, 280   

                                            Посилання Coordinates  Latitude  \
1   https://dozorro.org/tenderUA-2018-09-14-000391...        None       nan   
19  https://dozorro.org/tenderUA-2018-02-27-001070...        None       nan   
20  https://dozorro.org/tenderUA-2018-07-06-000592...        None       nan   
21  https://dozorro.org/tenderUA-2018-07-12-000343...        None       nan   
22  https://dozorro.org/tenderUA-2018-01-18-000738...        None       nan   
23  https://dozorro.org/tenderUA-2018-01-26-002398...        None       nan   
24  https://dozorro.org/tenderUA-2018-02-01-000717...        None       nan   
28  https://dozorro.org/tenderUA-2018-07-13-001260...        None       nan   

    Longitude  
1         nan  
19        nan  
20        nan  
21        nan  
22        nan  
23        nan  
24        nan  
28        nan

In [220]:
from geopy.extra.rate_limiter import RateLimiter
geocode = RateLimiter(geolocator.geocode, min_delay_seconds=1)

df_map_error_address["Coordinates"]=df_map_error_address["Адреса закладу"].apply(geocode)

df_map_error_address["Latitude"]=df_map_error_address["Coordinates"].apply(lambda x: x.latitude if x != None else None)
df_map_error_address["Longitude"]=df_map_error_address["Coordinates"].apply(lambda x: x.longitude if x != None else None)

C:\Users\user1\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  after removing the cwd from sys.path.
C:\Users\user1\Anaconda3\lib\site-packages\ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\user1\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org

In [215]:
x = df_map_error_address['Адреса закладу'].tolist()
names = []

for item in x:
    d={}
    a = geolocator.geocode(item, exactly_one=True, timeout=60)
    try:
        df_map_error_address["Latitude"] = a.latitude
    except:
        pass
    try:
        df_map_error_address["Longitude"] = a.longitude
    except:
        pass
    time.sleep(2)
    names.append(d)

C:\Users\user1\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\user1\Anaconda3\lib\site-packages\ipykernel_launcher.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if sys.path[0] == '':


In [217]:
df_map_error_address

Заклад   Витрачена сума  \
1                                     Черкаси, Вулкан    68,316.000000   
19                                   Черкаси, ДЮСШ №1   141,594.200000   
20                          Черкаси, Музична школа №1 1,613,234.720000   
21                          Черкаси, Музична школа №3   581,440.000000   
22                     Черкаси, Перша міська гімназія 5,158,431.520000   
23  Черкаси, Центр дитячої та юнацької творчості м...   716,249.800000   
24  Черкаси, Черкаська дитяча художня школа ім. Да...   245,415.200000   
28  Черкаси, Черкаський фізико-математичний ліцей ... 3,066,886.090000   

                                      Адреса закладу  \
1                       Черкаси, вул. Ярославська, 6   
19                           Черкаси, Смілянська, 94   
20                       Черкаси, Б.Вишневецького,33   
21  Черкаси, ВУЛИЦЯ В'ЯЧЕСЛАВА ЧОРНОВОЛА, будинок 10   
22                   Черкаси, вул. Святотроїцька, 68   
23                 Черкаси, вул. Смілянська, буд. 33   
24                           Черкаси, Хрещатик,  214   
28                     Черкаси, вул. Благовісна, 280   

                                            Посилання  \
1   https://dozorro.org/tenderUA-2018-09-14-000391...   
19  https://dozorro.org/tenderUA-2018-02-27-001070...   
20  https://dozorro.org/tenderUA-2018-07-06-000592...   
21  https://dozorro.org/tenderUA-2018-07-12-000343...   
22  https://dozorro.org/tenderUA-2018-01-18-000738...   
23  https://dozorro.org/tenderUA-2018-01-26-002398...   
24  https://dozorro.org/tenderUA-2018-02-01-000717...   
28  https://dozorro.org/tenderUA-2018-07-13-001260...   

                                          Coordinates  Latitude  Longitude  
1                                                None 49.628059  31.619419  
19                                               None 49.628059  31.619419  
20  (33, Байди Вишневецького вулиця, Черкаси, Сосн... 49.628059  31.619419  
21                                               None 49.628059  31.619419  
22                                               None 49.628059  31.619419  
23                                               None 49.628059  31.619419  
24  (Хрещатик, Черкаський район, Черкаська область... 49.628059  31.619419  
28                                               None 49.628059  31.619419

In [ ]:
df_map_errorfix1 = df_map_error1[df_map_error1['Latitude'].notnull()]

df_map_error1.shape

df_map_errorfix1

In [143]:
df_map_errorfix2 = df_map_error[df_map_error['Адреса закладу'].isnull()]
df_map_errorfix2['Latitude'] = ''
df_map_errorfix2['Longitude'] = ''
df_map_errorfix2['Coordinates'] = ''
#df_map_errorfix1 = df_map_errorfix1[df_map_error['Latitude'].notnull()]

df_map_errorfix2

Empty DataFrame
Columns: [Заклад, Витрачена сума, Адреса закладу, Посилання, Coordinates, Latitude, Longitude]
Index: []

In [221]:
df_map_errorfix2["Адреса закладу"]=df_map_error["Заклад"]

df_map_errorfix2['Адреса закладу']=df_map_errorfix2['Адреса закладу'].str.replace('ДНЗ', 'дитячий садок')

df_map_errorfix2.to_excel('CherkasyError2.xls', index = True, header=True)

In [123]:
df_map_errorfix2 = pd.read_excel('CherkasyError2.xls', encoding='UTF-8')
df_map_errorfix2.drop(df_map_errorfix2.columns[df_map_errorfix2.columns.str.contains('unnamed',case = False)],axis = 1, inplace = True)
df_map_errorfix2.head(2)

Заклад  Витрачена сума              Адреса закладу  \
0  Черкаси, ДНЗ №22  118,890.000000  Черкаси, дитячий садок №22   
1  Черкаси, ДНЗ №25  197,411.000000  Черкаси, дитячий садок №25   

                                          Посилання           Coordinates  \
0  https://dozorro.org/tenderUA-2018-12-21-000170-b  49.423767, 32.090116   
1  https://dozorro.org/tenderUA-2018-12-21-000469-c  49.435240, 31.999330   

   Latitude  Longitude  
0 49.423767  32.090116  
1 49.435240  31.999330

df_map_error['Заклад'][4] #= '214, вулиця Хрещатик, Черкаси, Черкаська область, 18000'

In [132]:
frames = [df_map_no_error, df_map_error, df_map_errorfix2]

df_full = pd.concat(frames, ignore_index=True)

In [133]:
df_full

Заклад    Витрачена сума  \
0                                   Черкаси, Берегиня  3,589,376.000000   
1                                    Черкаси, ДНЗ №39    357,443.600000   
2                                    Черкаси, ДНЗ №45    921,979.400000   
3                                    Черкаси, ДНЗ №57  1,117,106.960000   
4                                    Черкаси, ДНЗ №59  1,051,721.260000   
5                                    Черкаси, ДНЗ №81  1,233,499.200000   
6                                    Черкаси, ДНЗ №91  1,052,301.090000   
7            Черкаси, Черкаська дитяча школа мистецтв  3,185,948.880000   
8   Черкаси, Черкаський професійний автодорожній л...  1,508,684.330000   
9               Черкаси, Черкаський професійний ліцей  2,577,956.840000   
10                                 Черкаси, Школа №10  2,509,415.430000   
11                                 Черкаси, Школа №11  2,404,258.230000   
12                                 Черкаси, Школа №12  3,255,307.130000   
13                                 Черкаси, Школа №13  4,737,529.320000   
14                                 Черкаси, Школа №17  8,380,343.860000   
15                                 Черкаси, Школа №18  6,198,063.140000   
16                                 Черкаси, Школа №19    491,465.300000   
17                                  Черкаси, Школа №2  2,457,121.980000   
18                                 Черкаси, Школа №20  3,973,798.320000   
19                                 Черкаси, Школа №21  3,084,428.210000   
20                                 Черкаси, Школа №22  3,494,725.020000   
21                                 Черкаси, Школа №24  4,789,368.050000   
22                                 Черкаси, Школа №26  7,950,985.020000   
23                                 Черкаси, Школа №27  4,589,153.320000   
24                                 Черкаси, Школа №28    348,390.830000   
25                                  Черкаси, Школа №3  5,031,781.140000   
26                                 Черкаси, Школа №30  4,578,438.150000   
27                                 Черкаси, Школа №31  7,864,100.370000   
28                                 Черкаси, Школа №32  8,659,487.420000   
29                                 Черкаси, Школа №33  6,872,410.400000   
30                                 Черкаси, Школа №34  2,473,302.100000   
31                                  Черкаси, Школа №4  2,495,789.430000   
32                                  Черкаси, Школа №5  2,410,312.520000   
33                                  Черкаси, Школа №6  4,355,353.240000   
34                                  Черкаси, Школа №7  2,682,962.030000   
35                                  Черкаси, Школа №8  2,586,596.730000   
36                                  Черкаси, Школа №9 11,849,415.220000   
37                                    Черкаси, Вулкан     68,316.000000   
38                                   Черкаси, ДЮСШ №1    141,594.200000   
39                          Черкаси, Музична школа №1  1,613,234.720000   
40                          Черкаси, Музична школа №3    581,440.000000   
41                     Черкаси, Перша міська гімназія  5,158,431.520000   
42  Черкаси, Центр дитячої та юнацької творчості м...    716,249.800000   
43  Черкаси, Черкаська дитяча художня школа ім. Да...    245,415.200000   
44  Черкаси, Черкаський фізико-математичний ліцей ...  3,066,886.090000   
45                                   Черкаси, ДНЗ №22    118,890.000000   
46                                   Черкаси, ДНЗ №25    197,411.000000   
47                                   Черкаси, ДНЗ №27    171,445.220000   
48                                   Черкаси, ДНЗ №29  2,473,554.000000   
49                                   Черкаси, ДНЗ №30    295,673.000000   
50                                   Черкаси, ДНЗ №37     68,901.000000   
51                                   Черкаси, ДНЗ №38    479,047.000000   
52                                   Черкаси, ДНЗ №41    572,465.000000 

In [134]:
df_map_pivot

Заклад    Витрачена сума  \
0                                   Черкаси, Берегиня  3,589,376.000000   
1                                     Черкаси, Вулкан     68,316.000000   
2                                    Черкаси, ДНЗ №22    118,890.000000   
3                                    Черкаси, ДНЗ №25    197,411.000000   
4                                    Черкаси, ДНЗ №27    171,445.220000   
5                                    Черкаси, ДНЗ №29  2,473,554.000000   
6                                    Черкаси, ДНЗ №30    295,673.000000   
7                                    Черкаси, ДНЗ №37     68,901.000000   
8                                    Черкаси, ДНЗ №38    479,047.000000   
9                                    Черкаси, ДНЗ №39    357,443.600000   
10                                   Черкаси, ДНЗ №41    572,465.000000   
11                                   Черкаси, ДНЗ №45    921,979.400000   
12                                   Черкаси, ДНЗ №57  1,117,106.960000   
13                                   Черкаси, ДНЗ №59  1,051,721.260000   
14                                    Черкаси, ДНЗ №7    762,226.300000   
15                                   Черкаси, ДНЗ №70    912,793.340000   
16                                   Черкаси, ДНЗ №73  2,244,335.200000   
17                                   Черкаси, ДНЗ №81  1,233,499.200000   
18                                   Черкаси, ДНЗ №91  1,052,301.090000   
19                                   Черкаси, ДЮСШ №1    141,594.200000   
20                          Черкаси, Музична школа №1  1,613,234.720000   
21                          Черкаси, Музична школа №3    581,440.000000   
22                     Черкаси, Перша міська гімназія  5,158,431.520000   
23  Черкаси, Центр дитячої та юнацької творчості м...    716,249.800000   
24  Черкаси, Черкаська дитяча художня школа ім. Да...    245,415.200000   
25           Черкаси, Черкаська дитяча школа мистецтв  3,185,948.880000   
26  Черкаси, Черкаський професійний автодорожній л...  1,508,684.330000   
27              Черкаси, Черкаський професійний ліцей  2,577,956.840000   
28  Черкаси, Черкаський фізико-математичний ліцей ...  3,066,886.090000   
29                                 Черкаси, Школа №10  2,509,415.430000   
30                                 Черкаси, Школа №11  2,404,258.230000   
31                                 Черкаси, Школа №12  3,255,307.130000   
32                                 Черкаси, Школа №13  4,737,529.320000   
33                                 Черкаси, Школа №17  8,380,343.860000   
34                                 Черкаси, Школа №18  6,198,063.140000   
35                                 Черкаси, Школа №19    491,465.300000   
36                                  Черкаси, Школа №2  2,457,121.980000   
37                                 Черкаси, Школа №20  3,973,798.320000   
38                                 Черкаси, Школа №21  3,084,428.210000   
39                                 Черкаси, Школа №22  3,494,725.020000   
40                                 Черкаси, Школа №24  4,789,368.050000   
41                                 Черкаси, Школа №26  7,950,985.020000   
42                                 Черкаси, Школа №27  4,589,153.320000   
43                                 Черкаси, Школа №28    348,390.830000   
44                                  Черкаси, Школа №3  5,031,781.140000   
45                                 Черкаси, Школа №30  4,578,438.150000   
46                                 Черкаси, Школа №31  7,864,100.370000   
47                                 Черкаси, Школа №32  8,659,487.420000   
48                                 Черкаси, Школа №33  6,872,410.400000   
49                                 Черкаси, Школа №34  2,473,302.100000   
50                                  Черкаси, Школа №4  2,495,789.430000   
51                                  Черкаси, Школа №5  2,410,312.520000   
52                                  Черкаси, Школа №6  4,355,353.240000 